src/v4.ts

```diff
@@ -0,0 +1,1201 @@
+        }
+      } catch (err) {
+        return false;
+      } finally {
+        this.metric('canOpen');
+      }
+      return true;
+    });
+  }
+  // This mostly happens after removing file. Checks if directory still exists.
+  // TODO: either use real readdir (with updating information) or remove this
+  private async canOpenDir(dir: Path) {
+    dir = sysPath.resolve(dir);
+    return this.ioLimit(async () => {
+      const items = this.getWatchedDir(dir);
+      if (items.size > 0) return;
+      try {
+        await readdir(dir);
+      } catch (err) {
+        this.remove(sysPath.dirname(dir), sysPath.basename(dir));
+      }
+    });
+  }
+  // /IO
+  // Utils
+  /**
+   * Provides directory tracking objects
+   */
+  private getWatchedDir(directory: string): Set<string> {
+    const dir = sysPath.resolve(directory);
+    if (!this.watched.has(dir)) this.watched.set(dir, new Set());
+    return this.watched.get(dir);
+  }
+  private normalizePath(path: Path) {
+    const { cwd } = this.options;
+    path = sysPath.normalize(path);
+    // TODO: do we really need that? only thing it does is using '//' instead of '\\' for network shares
+    // in windows. Path normalize already strips '//' in windows.
+    // > If SLASH_SLASH occurs at the beginning of path, it is not replaced
+    // >    because "//StoragePC/DrivePool/Movies" is a valid network path
+    path = path.replace(/\\/g, '/');
+    let prepend = false;
+    if (path.startsWith('//')) prepend = true;
+    const DOUBLE_SLASH_RE = /\/\//;
+    while (path.match(DOUBLE_SLASH_RE)) path = path.replace(DOUBLE_SLASH_RE, '/');
+    if (prepend) path = '/' + path;
+    // NOTE: join will undo all normalization
+    if (cwd) path = sysPath.isAbsolute(path) ? path : sysPath.join(cwd, path);
+    return path;
+  }
+  private normalizePaths(paths: Path | Path[]) {
+    // TODO: do we really need flatten here?
+    paths = flatten(arrify(paths));
+    if (!paths.every((p) => typeof p === 'string'))
+      throw new TypeError(`Non-string provided as watch path: ${paths}`);
+    return paths.map((i) => this.normalizePath(i));
+  }
+  /**
+   * Helper utility for throttling
+   * @param actionType type being throttled
+   * @param path being acted upon
+   * @param ms duration to suppress duplicate actions
+   * @returns tracking object or false if action should be suppressed
+   */
+  private throttle(actionType: ThrottleType, path: Path, ms: number) {
+    // NOTE: this is only used correctly in readdir for now.
+    // How it should work:
+    // - we process some event
+    // - same event happens in parallel multiple times
+    // - when we processed first event, we look at last throttled event
+    // - start processing it
+    // How it works now (except readdir):
+    // - we process first event (first change)
+    // - there is multiple parallel changes which we throw away
+    // - all changes which happened when we processed first event is lost
```

Comment:
fixing this would probably also solve the windows bug

just a recap of what that is/was:

writing a file in windows often fires two events, while it fires one on linux/mac